### PR TITLE
kube-scheduler remove non-csi volumelimit plugins

### LINKED ISSUE #82375: [scheduler] Remove deprecated EBS/GCE/Azure/Cinder predicates
/sig scheduling
/priority important-longterm
/assign

We should remove the EBS/GCE/Azure/Cinder predicates from scheduler after CSIMigration goes GA (probably in 1.18).

https://github.com/kubernetes/kubernetes/blob/8c537b46939beed3713356f71449a8ad6829b725/pkg/scheduler/algorithm/predicates/predicates.go#L80-L95

> Since we are in process of migrating in-tree volume plugins to CSI, there is already a CSI predicate available which handles all volume plugins and hence when using CSI volumes - the cloudprovider specific predicates are not required. We will be able to truly remove these predicates once CSI migration goes GA but we will have a translation layer as well, which would take care of handling in-tree volume plugins even if you are not using CSI. https://github.com/kubernetes/kubernetes/pull/74544#issuecomment-467531169

#### Reference

+ https://github.com/kubernetes/kubernetes/pull/74544#issuecomment-467531169
+ https://github.com/kubernetes/enhancements/issues/625

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

See https://github.com/kubernetes/kubernetes/pull/123970

#### Which issue(s) this PR fixes:

Fixes #82375

#### Does this PR introduce a user-facing change?

```release-note
kube-scheduler removed the following plugins: 
- AzureDiskLimits 
- CinderLimits
- EBSLimits
- GCEPDLimits
Because the corresponding CSI driver reports how many volumes a node can handle in NodeGetInfoResponse, the kubelet stores this limit in CSINode and the scheduler then knows the driver's limit on the node.
ACTION REQUIRED: Remove plugins AzureDiskLimits, CinderLimits, EBSLimits and GCEPDLimits if you explicitly enabled them in the scheduler config.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/names/names.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/names/names.go:36]: This change could be a breaking change for users who explicitly uses those plugins in the scheduler config.

@kubernetes/sig-scheduling-approvers How do we usually deal with plugin removal? Just removing while mentioning in the release note? Or do we have to follow a deprecation process or something?

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: Those plugins have been deprecated by https://github.com/kubernetes/kubernetes/pull/74544 in 1.14.

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: BTW, I update this PR's description.

Fixes https://github.com/kubernetes/kubernetes/issues/82375

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/names/names.go:36]: Got it, can you add more information on the release note? e.g., do users of those plugins have to migrate to something else? Or no action is required at all?

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: I think there's no required action caused by this PR. Because those volume types are always translated to the CSI type. If the node status contains volume limitation info, the csi scheduler plugin will use it even though the csinode doesn't include this info.

FYI: https://github.com/kubernetes/kubernetes/blob/6673e7a93d8f0e81343309350e3ec7e7afc7fd73/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L406-L409

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: Oh, I forgot that I also changed the CSI plugin in this PR. It wouldn't read limits from the node object anymore. If a storage vendor wants to use the limits, they have to create manually the CSI driver object when deploying the driver. It should be a standard way to deploy a CSI driver if limits are needed. So I want to know whether I need to mention this in the release note. If so, I will add it in the release note. cc @sanposhiho what do you think?

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/names/names.go:36]: > I think there's no other required action caused by this PR. Because those volume types are always translated to the CSI type. If the node status contains volume limitation info, the csi scheduler plugin will use it even though the csinode doesn't include this info.

OK, then please mention it; e.g., `Because XXXX, no action is required for users for this removal. Please just remove those plugins if you explicitly use those plugins in the scheduler config.`

>  I want to know whether I need to mention this in the release note. If so, I will add it in the release note. cc @sanposhiho what do you think?

I believe that's worth enough to be mentioned in the release note, but defer to someone from sig-storage as I don't have much context around that.

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: cc @jsafrane

Comment (jsafrane) [reply pkg/scheduler/framework/plugins/names/names.go:36]: I agree with @carlory, Azure, GCE, AWS and Cinder Limits, using data from Node objects are not used any longer. These volume plugins were migrated to CSI and CSINode object is used instead. These plugins should not do anything in v1.30.

What does kube-scheduler process do with unknown plugins specified in its config file? If it exits, then 
this PR could break clusters who explicitly enabled/disabled the plugins there. 

> If a storage vendor wants to use the limits, they have to create manually the CSI driver object when deploying the driver. It should be a standard way to deploy a CSI driver if limits are needed.

It's actually automatic. A CSI driver reports how much volumes can a node handle in [NodeGetInfoResponse](https://github.com/container-storage-interface/spec/blob/2696773158b970c65107610d8f3ef4302546eca0/csi.proto#L1689) and kubelet stores this limit in CSINode. Scheduler then knows the driver's limit on the node. And each node can have a different limit.

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/names/names.go:36]: > What does kube-scheduler process do with unknown plugins specified in its config file? If it exits, then
this PR could break clusters who explicitly enabled/disabled the plugins there.

It exits -
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework.go#L628

But, according to https://github.com/kubernetes/kubernetes/pull/74544 @carlory shew, those plugins are already marked as deprecated. So, it should be OK to just remove them as long as we mention it in the release note, I believe.

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: @sanposhiho the release note is updated.

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/names/names.go:36]: @jsafrane Please take a look at the release note from sig-storage pov too.

**Comment by sanposhiho** [comment]
@kubernetes/sig-scheduling-approvers PTAL for the second review path.

File: pkg/scheduler/framework/plugins/names/names.go
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/names/names.go:36]: Can we remove these plugins from default plugins first and make them invalid the same time: https://github.com/kubernetes/kubernetes/blob/9c9a0e3ba57cb103d651e77f1b9283673c4d0cd7/pkg/scheduler/apis/config/validation/validation.go#L101-L106 This is how we used to deprecate plugins, then next release let's remove the total logics together, I know these plugins are already marked as deprecated for releases, but let's be less surprised to users as remove plugins from default ones directly.

If so let's mention that **Those plugins are removed from default plugins as they're already marked as deprecated since 1.14 and will be totally removed in v1.31, also mention the replacement we have**

Or we're blocked by these plugins like schedulingHint?

Comment (carlory) [reply pkg/scheduler/framework/plugins/names/names.go:36]: /hold 
merge it in 1.32

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/names/names.go:36]: /un-lgtm 
for now

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/names/names.go:36]: /remove-lgtm

**Comment by jsafrane** [comment]
/lgtm
From storage point of view. Sadly, I can't help with approvals.

**Comment by sanposhiho** [comment]
/approve

Per the discussion [here](https://github.com/kubernetes/kubernetes/pull/124003#discussion_r1574628509), we should be able to merge it after the code freeze of v1.31.

cc @kerthcet for the cross-check.

**Comment by sanposhiho** [comment]
Well, I guess SIG-Scheduling or SIG-Storage should own ↓ actually.. Currently no owner file and thus `/tests/` owner owns them.
https://github.com/kubernetes/kubernetes/tree/master/test/integration/volumescheduling

**Comment by carlory** [comment]
/cc @saad-ali @msau42 @aojea

Can you have a look?

**Review by pohly — APPROVED** [review]
/approve

**Comment by alculquicondor** [comment]
> no action is required for users for this removal.

Is this 100% true?

If a user was explicitly enabling any of these plugins, they need to stop doing that and make sure not enable the CSI plugins.

**Comment by carlory** [comment]
@alculquicondor scheduler deprecated non-CSI volume limit plugins in 1.31.  https://github.com/kubernetes/kubernetes/pull/124500

**Comment by sanposhiho** [comment]
> no action is required for users for this removal. 
Please just remove those plugins if you explicitly use those plugins in the scheduler config.

@carlory I think this part in the release note is contradictory. 
better to remove "no action is required for users for this removal."

**Comment by carlory** [comment]
@sanposhiho removed

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
kube-scheduler removed the following plugins: 
- AzureDiskLimits 
- CinderLimits
- EBSLimits
- GCEPDLimits
Because the corresponding CSI driver reports how many volumes a node can handle in NodeGetInfoResponse, the kubelet stores this limit in CSINode and the scheduler then knows the driver's limit on the node.
ACTION REQUIRED: Remove plugins AzureDiskLimits, CinderLimits, EBSLimits and GCEPDLimits if you explicitly enabled them in the scheduler config.
```

**Review by hashim21223445 — APPROVED** [review]

**Pull request merged**
